fem/fe/fe_rbf.hpp

```diff
@@ -0,0 +1,556 @@
+   mutable DenseMatrix ddr_scr;
+   mutable int cInd[3][2];
+#endif
+   bool isCompact; // Is the RBF with the given h compact?
+   int dimPoints[3];
+   int numPointsD; // Number of points across the element in each D
+   double delta; // Distance between points
+   double h; // Shape parameter, approx number of points in 1d support radius
+   double hPhys; // Shape parameter times distance between points times HNorm
+   double hPhysInv; // Inverse hPhys
+   double radPhys; // Radius adjusted by h
+   double faceFactor; // 1.0 = points end on face, 0.0 = points end at dx/2 from face
+   const RBFFunction *rbf;
+   const DistanceMetric *distance;
+   void InitializeGeometry();
+
+   // Get the dimensionless distance from x to the center of the RBF indexed i
+   virtual void DistanceVec(const int i,
+                            const Vector &x,
+                            Vector &y) const;
+
+public:
+   RBFFiniteElement(const int D,
+                    const int numPointsD,
+                    const int rbfType,
+                    const int distNorm,
+                    const int intOrder,
+                    const double h,
+                    const double faceFactor);
+   virtual ~RBFFiniteElement() { delete rbf; delete distance; }
+
+   virtual bool TensorIndexed() const { return true; }
+   virtual void GetCompactIndices(const Vector &ip,
+                                  int (&indices)[3][2]) const;
+   virtual void GetGlobalIndices(const Vector &ip,
+                                 int (&indices)[3][2]) const;
+   virtual void GetTensorIndices(const Vector &ip,
+                                 int (&indices)[3][2]) const;
+   virtual void GetTensorNumPoints(int (&tNumPoints)[3]) const
+   {
+      tNumPoints[0] = dimPoints[0];
+      tNumPoints[1] = dimPoints[1];
+      tNumPoints[2] = dimPoints[2];
+   }
+
+   virtual bool IsCompact() const { return isCompact; }
+   virtual const RBFFunction *Kernel() const { return rbf; }
+
+   virtual void CalcShape(const IntegrationPoint &ip,
+                          Vector &shape) const;
+   virtual void CalcDShape(const IntegrationPoint &ip,
+                           DenseMatrix &dshape) const;
+   virtual void CalcHessian(const IntegrationPoint &ip,
+                            DenseMatrix &hess) const;
+};
+
+/* Reproducing kernel finite element, which includes polynomial corrections
+   to the standard radial basis function finite element to guarantee a
+   chosen order of accuracy */
+class RKFiniteElement : public KernelFiniteElement
+{
+private:
+#ifndef MFEM_THREAD_SAFE
+   mutable double f_scr;
+   mutable Vector x_scr, y_scr, g_scr, c_scr, s_scr, p_scr, df_scr;
+   mutable DenseMatrix q_scr, dq_scr, M_scr;
+   mutable Vector dc_scr[3], dp_scr[3];
+   mutable DenseMatrix dM_scr[3];
+   mutable DenseMatrixInverse Minv_scr;
+   mutable int cInd[3][2];
+   mutable int dimPoints[3];
+#endif
+   int polyOrd, numPoly, numPoly1d;
+   KernelFiniteElement *baseFE;
```

Comment:
Don't you know it's always a `RBFFiniteElement`?

Reply:
For now, yes. Since we aren't using any functions from RBFFiniteElement that aren't in KernelFiniteElement, it made sense to keep it open to additions in the future, for example if we eventually try a RBF finite element type that is a tensor product of 1D functions.

Reply:
It's a private member, you can make it abstract in the future if it's needed.

Reply:
I think my main reason to make it a KernelFiniteElement was to only use the functions there and avoid using the RBFFiniteElement functions. Then if we want to do a different class with RK, we won't have to go back and check that it doesn't use any functions only defined in RBFFiniteElement.